src/net.cpp

```diff
@@ -2539,16 +2539,15 @@ std::vector<CAddress> CConnman::GetAddresses()
     return addresses;
 }
 
-std::vector<CAddress> CConnman::GetAddresses(Network requestor_network, CAddress local_socket_address)
+const std::vector<CAddress>& CConnman::GetAddresses(Network requestor_network, CAddress local_socket_address)
 {
     auto local_socket_bytes = local_socket_address.GetAddrBytes();
     uint64_t cache_id = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE)
         .Write(requestor_network)
         .Write(local_socket_bytes.data(), local_socket_bytes.size())
         .Finalize();
     const auto current_time = GetTime<std::chrono::microseconds>();
-    if (m_addr_response_caches.find(cache_id) == m_addr_response_caches.end() ||
-        m_addr_response_caches[cache_id].m_update_addr_response < current_time) {
+    if (m_addr_response_caches.count(cache_id) == 0 || m_addr_response_caches[cache_id].m_cache_expiration_time < current_time) {
```

Comment:
96cd22cd3204b35b9f863ada93a99dcfac178d13

You have 4 `m_addr_response_caches` lookups. How about (untested):
```
auto r = m_addr_response_caches.emplace(cache_id, {}};
if (e.second || e.first->m_cache_expiration_time < current_time) {
    r.first->m_addrs_response_cache = GetAddresses());
    r.first->m_cache_expiration_time = ...;
}
return r.first->m_addrs_response_cache;

Reply:
<details>
<summary>diff</summary>

```diff
diff --git i/src/net.cpp w/src/net.cpp
index eb462d108..169c4cd1c 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2544,14 +2544,16 @@ const std::vector<CAddress>& CConnman::GetAddresses(Network requestor_network, C
     auto local_socket_bytes = local_socket_address.GetAddrBytes();
     uint64_t cache_id = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE)
         .Write(requestor_network)
         .Write(local_socket_bytes.data(), local_socket_bytes.size())
         .Finalize();
     const auto current_time = GetTime<std::chrono::microseconds>();
-    if (m_addr_response_caches.count(cache_id) == 0 || m_addr_response_caches[cache_id].m_cache_expiration_time < current_time) {
-        m_addr_response_caches[cache_id].m_addrs_response_cache = GetAddresses();
+    auto r = m_addr_response_caches.emplace(cache_id, CachedAddrResponse{});
+    auto& cache = r.first->second;
+    if (cache.m_cache_expiration_time < current_time) { // If emplace() added new one it has expiration 0.
+        cache.m_addrs_response_cache = GetAddresses();
 
         // Choosing a proper cache lifetime a trade-off between the privacy leak minimization
         // and the usefulness of ADDR responses to honest users.
         //
         // Longer cache lifetime makes it more difficult for an attacker to scrape
         // enough AddrMan data to maliciously infer something useful.
@@ -2571,15 +2573,15 @@ const std::vector<CAddress>& CConnman::GetAddresses(Network requestor_network, C
         // in the ADDR response are no longer active.
         //
         // However, the churn in the network is known to be rather low. Since we consider
         // nodes to be "terrible" (see IsTerrible()) if the timestamps are older than 30 days,
         // max. 24 hours of "penalty" due to cache shouldn't make any meaningful difference
         // in terms of the freshness of the response.
-        m_addr_response_caches[cache_id].m_cache_expiration_time = current_time + std::chrono::hours(21) + GetRandMillis(std::chrono::hours(6));
+        cache.m_cache_expiration_time = current_time + std::chrono::hours(21) + GetRandMillis(std::chrono::hours(6));
     }
-    return m_addr_response_caches[cache_id].m_addrs_response_cache;
+    return cache.m_addrs_response_cache;
 }
 
 bool CConnman::AddNode(const std::string& strNode)
 {
     LOCK(cs_vAddedNodes);
     for (const std::string& it : vAddedNodes) {
```
</details>